tests/cephadm/test_custom_container.py

```diff
@@ -0,0 +1,156 @@
+
+    # Setup ssh key to all nodes
+    setup_ssh_keys(installer, ceph_nodes)
+
+    # Get registry credentials
+    registry, reg_username, reg_password = get_registry_details(config)
+
+    # Setup a private registry on installer node
+    setup_private_container_registry(
+        installer=installer,
+        nodes=ceph_nodes,
+        registry=registry,
+        reg_username=reg_username,
+        reg_password=reg_password,
+        build_type=rhbuild,
+        private_reg_username=PRIVATE_REG_USER,
+        private_reg_password=PRIVATE_REG_PASS,
+        docker_reg_image=docker_reg_image,
+        images=dashboard_images,
+    )
+
+    # Get private registry catalog
+    out = get_private_registry_image(
+        installer, username=PRIVATE_REG_USER, password=PRIVATE_REG_PASS
+    )
+    repos = out.get("repositories")
+
+    # Pull images to nodes
+    for ceph in ceph_nodes:
+        Registry(ceph).login(
+            registry=f"{installer.hostname}:{PORT}",
+            username=PRIVATE_REG_USER,
+            password=PRIVATE_REG_PASS,
+        )
+        for img in repos:
+            Container(ceph).pull(image=f"{installer.hostname}:{PORT}/{img}")
+
+    for config_key, image in registry_image.items():
+        img_reg = image.replace(f"{registry}/", "")
+
+        # Set mgr config for dashboard images
+        conf = {
+            "key": "mgr",
+            "value": f"mgr/cephadm/{config_key} {installer.hostname}:{PORT}/{img_reg}",
+        }
+        cephadm.ceph.config.set(**conf)
+
+        # Get mgr config for dashboard images
+        conf = {
+            "who": "mgr",
+            "key": f"mgr/cephadm/{config_key}",
+        }
+        out = cephadm.ceph.config.get(**conf)[0]
+
+        # Validate if dashboard images are updated correctly
+        if installer.hostname not in out:
+            raise ClusterCustomContainerError(
+                f"Failed to set custom image mgr/cephadm/{config_key}"
+            )
+
+    for config_key, _ in registry_image.items():
+        service = config_key.replace("container_image_", "")
+        service = service.replace("_", "-")
+
+        # Redeploy dashboard services
+        conf = {"service": service}
+        cephadm.ceph.orch.redeploy(**conf)
+
+        timeout, interval = 30, 5
+        for w in WaitUntil(timeout=timeout, interval=interval):
+            conf = {"refresh": True}
+            cephadm.ceph.orch.ps(**conf)
+
+            # Verify dashboard services updated with private images
+            service_info = loads(
+                cephadm.ceph.orch.ps(service_name=service, format="json")
+            )
+            image = [c.get("container_image_name") for c in service_info][0]
+            status = [c.get("status_desc") for c in service_info][0]
+            if status == "running" and installer.hostname in image:
+                log.info(
+                    f"{service} has been re-deployed with custom-image successfully"
+                )
+                break
```

Comment:
Add error message for retry

Reply:
Vaibhav, we have 
        if w.expired:
            raise ClusterCustomContainerError(
                f"Monitor service deployment failed for {service}"
            )
If retry will fail, do we need to add any other error message?

Reply:
Updated the info message for retry

Reply:
use -
```
            log.info(f"Service '{service}' is not in expected '{status}' state, retry after {interval} sec")
```